### PR TITLE
Remove mmap.extensions setting

### PR DESCRIPTION
**Pull request by dzane17**
### Description
Removes `index.store.hybrid.mmap.extensions` setting in OS 3.0.
Follow up to https://github.com/opensearch-project/OpenSearch/pull/8508

### Background
Currently OpenSearch code contains an explicit list of file extensions which load using mmap from hybridfs. Other file extensions default to nio. This PR flips the logic to keep a list of nio file extensions while all others default to mmap. This will prevent any future regressions in case Lucene adds new segment file type that should be using mmap.

### Related Issues
Resolves https://github.com/opensearch-project/OpenSearch/issues/8297

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/index/IndexModule.java
Comment (reta) [inline server/src/main/java/org/opensearch/index/IndexModule.java:165]: @andrross @dblock @nknize seems like we are really rushing the removal here: the feature will be deprecated in 2.10.0 but already removed in 3.0.0, it looks to be very short notice ...

File: CHANGELOG.md
Comment (andrross) [inline CHANGELOG.md:53]: This can go in the `# Removed` section

File: server/src/main/java/org/opensearch/index/IndexModule.java
Comment (andrross) [reply server/src/main/java/org/opensearch/index/IndexModule.java:165]: I think semver policy technically allows for deprecating something in any minor version and removing in the next major, even if those end up being consecutive releases. Is that wrong?

For a widely used feature that requires complicated changes changes to move away from then for sure we'd want to give more runway than that. For an expert setting like this, I'm more comfortable with the faster removal schedule. We can carry this for another major version though if you think this will be too disruptive.

File: CHANGELOG.md
Comment (dzane17) [reply CHANGELOG.md]: moved it

File: server/src/main/java/org/opensearch/index/IndexModule.java
Comment (reta) [reply server/src/main/java/org/opensearch/index/IndexModule.java:165]: > I think semver policy technically allows for deprecating something in any minor version and removing in the next major, even if those end up being consecutive releases. Is that wrong?

Not generally wrong, doing that in `2.0` would be 100% OK, in `2.10.0` raises the question how much time (till `3.0.0`) we give users to migrate

Comment (dzane17) [reply server/src/main/java/org/opensearch/index/IndexModule.java:165]: I am inclined to make this change in `3.0`. It is a major release so users can expect these type of changes.

Also even though we deprecate the setting in `2.10`, this does not guarantee users switch from using `mmap.extensions` to `nio.extensions`. Many will ignore the warning so keeping a longer migration window makes no difference.

Comment (reta) [reply server/src/main/java/org/opensearch/index/IndexModule.java:165]: @andrross have no objections to speed it up if you think so

> Many will ignore the warning so keeping a longer migration window makes no difference.

We don't know how widely this setting is being used, may be no one uses it, giving sufficient is the only thing we could do.

**Comment by andrross** [comment]
@dzane17 It looks like this change will break k-NN: https://github.com/opensearch-project/k-NN/blob/9702dd0ff19adc37844fd91201eb392fd2d2fc39/src/main/java/org/opensearch/knn/plugin/KNNPlugin.java#L359

Also the documentation will need to be updated: https://opensearch.org/docs/latest/search-plugins/knn/performance-tuning/#search-performance-tuning

We should probably do those things before we go forward with the removal.

File: server/src/test/java/org/opensearch/index/store/FsDirectoryFactoryTests.java
Comment (andrross) [inline server/src/test/java/org/opensearch/index/store/FsDirectoryFactoryTests.java:99]: Is this change related to the removal or is this something that should be backported?

**Comment by dzane17** [comment]
Documentation-website PR: https://github.com/opensearch-project/documentation-website/pull/4857

File: server/src/test/java/org/opensearch/index/store/FsDirectoryFactoryTests.java
Comment (dzane17) [reply server/src/test/java/org/opensearch/index/store/FsDirectoryFactoryTests.java:99]: This is part of removal, does not need to be backported.

**Comment by ticheng-aws** [comment]
Hi @dzane17, the PR is stalled. Is this being worked upon? Feel free to reach out to maintainers for further reviews.

**Comment by jainankitk** [comment]
@dzane17 - Can you address the merge the main branch and revive this PR? Good time to do that since we already thinking about 3.0.

**Comment by jainankitk** [comment]
> @andrross have no objections to speed it up if you think so
> 
> > Many will ignore the warning so keeping a longer migration window makes no difference.
> 
> We don't know how widely this setting is being used, may be no one uses it, giving sufficient is the only thing we could do.

@reta - Given we have had sometime since 2.10 was released and still thinking about 3.0, I am inclined to take this change for 3.0. Thoughts? cc: @andrross @dzane17

**Comment by andrross** [comment]
@jainankitk I'm fine with taking this change in 3.0.

**Comment by dzane17** [comment]
@jainankitk @andrross @reta I have rebased this PR. Can we please get it reviewed and merged?

File: CHANGELOG-3.0.md
Comment (jainankitk) [inline CHANGELOG-3.0.md:46]: Can be more verbose in change description:

`Remove mmap.extensions setting in favor of nio.extensions`

**Review by jainankitk — APPROVED** [review]
LGTM!

**Comment by reta** [comment]
> @reta - Given we have had sometime since 2.10 was released and still thinking about 3.0, I am inclined to take this change for 3.

+1 for 3.0, thanks @jainankitk

**Comment by jainankitk** [comment]
> > @reta - Given we have had sometime since 2.10 was released and still thinking about 3.0, I am inclined to take this change for 3.
> 
> +1 for 3.0, thanks @jainankitk

Thanks @reta.

**Comment by jainankitk** [comment]
@dzane17 - Please update the change description, and get this merged with help from @andrross or @reta or one of the other maintainers. Although I don't see any backport label, still ensure to not backport into 2.x.

File: CHANGELOG-3.0.md
Comment (dzane17) [reply CHANGELOG-3.0.md]: Updated changelog and commit message

**Comment by andrross** [comment]
@dzane17 @jainankitk This will break [k-NN](https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/plugin/KNNPlugin.java#L364-L375). We should update ensure k-NN gets updated before merging this change (either open an issue with a specific suggestion or create a PR directly).

**Comment by jainankitk** [comment]
> @dzane17 @jainankitk This will break [k-NN](https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/plugin/KNNPlugin.java#L364-L375). We should update ensure k-NN gets updated before merging this change (either open an issue with a specific suggestion or create a PR directly).

@navneet1v - Is [this](https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/plugin/KNNPlugin.java#L372) the only place KNN plugin is using this setting? Also, are there any file extensions that KNN needs to read as NIOFS? If not, MMAP is the new default unless overridden using NIOFS setting. cc: @vamshin

File: CHANGELOG-3.0.md
Comment (reta) [inline CHANGELOG-3.0.md:46]: Please:

```suggestion
- Remove `index.store.hybrid.mmap.extensions` setting in favor of `index.store.hybrid.nio.extensions` ([#9392](https://github.com/opensearch-project/OpenSearch/pull/9392))
```

Comment (dzane17) [reply CHANGELOG-3.0.md]: Thanks

**Comment by navneet1v** [comment]
> > @dzane17 @jainankitk This will break [k-NN](https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/plugin/KNNPlugin.java#L364-L375). We should update ensure k-NN gets updated before merging this change (either open an issue with a specific suggestion or create a PR directly).
> 
> @navneet1v - Is [this](https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/plugin/KNNPlugin.java#L372) the only place KNN plugin is using this setting? Also, are there any file extensions that KNN needs to read as NIOFS? If not, MMAP is the new default unless overridden using NIOFS setting. cc: @vamshin

KNN don't use any files as NIOFS.

**Review by reta — APPROVED** [review]

**Comment by jainankitk** [comment]
> KNN don't use any files as NIOFS.

@navneet1v - Great, thanks for confirming. Can you also create PR to remove the `additionalSettings` method from KNN plugin for 3.0? We can merge both the PRs together into mainline.

File: CHANGELOG-3.0.md
Comment (jainankitk) [reply CHANGELOG-3.0.md]: Thanks @reta, more verbose and precise, the better!

**Comment by reta** [comment]
@andrross LGTY?

**Review by andrross — APPROVED** [review]

**Pull request merged**
